### PR TITLE
schedule: implement dma_single_chan_domain

### PR DESCRIPTION
**Pull request by tlauda**
Adds implementation of DMA single channel scheduling domain.
This domain allows to schedule all the tasks on all the cores
on single DMA channel interrupt. This way we have only one
interrupt source and we don't need to worry about handling
many interrupts. We always select the channel with the lowest
period. This domain cannot be used with the DMAs, which require
to manually reload the next data to be transferred e.g. DW-DMA
on BYT or HSW.

Signed-off-by: Tomasz Lauda <tomasz.lauda@linux.intel.com>

### CODE REVIEW

**Comment by tlauda** [comment]
SOFCI TEST

File: src/schedule/dma_single_chan_domain.c [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+ */
+static int dma_single_chan_domain_irq_register(struct dma_chan_data *channel,
+					       struct dma_domain_data *data,
+					       void (*handler)(void *arg),
+					       void *arg)
+{
+	int irq = dma_chan_irq(channel->dma, channel->index);
+	int ret;
+
+	data->irq = interrupt_get_irq(irq, dma_irq_name(channel->dma));
```
Comment (paulstelian97) [inline src/schedule/dma_single_chan_domain.c]: As I mentioned on the already merged commit, the mere usage of a simple dma_irq_name macro which only considers the DMAC itself might cause problems on the i.MX platform, due to the interrupts for a DMAC being possibly spread among multiple cascaded controllers.

Comment (tlauda) [reply src/schedule/dma_single_chan_domain.c]: @paulstelian97 `dma_chan_irq` takes both dma and channel, so I don't see any reasons why not to extend `dma_irq_name` the same way. Feel free to do it.

Comment (paulstelian97) [reply src/schedule/dma_single_chan_domain.c]: Guess I'll do the change locally and introduce it together with the EDMA itself. Thanks for the suggestion.

**Comment by tlauda** [comment]
SOFCI TEST

File: src/schedule/dma_single_chan_domain.c [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+	bool channel_changed;	/* true if we needed to re-register */
+
+	/* data per core */
+	struct dma_domain_data data[PLATFORM_CORE_COUNT];
+};
+
+struct ll_schedule_domain_ops dma_single_chan_domain_ops;
+
+static void dma_single_chan_domain_enable(struct ll_schedule_domain *domain,
+					  int core);
```
Comment (jajanusz) [inline src/schedule/dma_single_chan_domain.c]: Is it misaligned or GH shows it incorrectly?

Comment (paulstelian97) [reply src/schedule/dma_single_chan_domain.c]: Looks like tabs are at it again. I think they're aligned.

File: src/schedule/ll_schedule.c [hunk 3]
```diff
@@ -128,20 +125,30 @@ static void schedule_ll_clients_reschedule(struct ll_schedule_data *sch)
...
 
+	spin_lock(sch->domain->lock);
+
+	/* clear domain only if all clients are done */
+	num_clients = atomic_sub(&sch->domain->num_clients, 1);
+	if (!num_clients)
+		domain_clear(sch->domain);
+
+	spin_unlock(sch->domain->lock);
+
```
Comment (jajanusz) [inline src/schedule/ll_schedule.c]: Do we really need lock if sub is atomic? Are you worried about domain_clear being executed in parallel? If subs are atomic there should be only 1 client that gets num_clients to 0 and calls domain clear, unless it's called also from some other place.

Comment (paulstelian97) [reply src/schedule/ll_schedule.c]: What about removing the last client but in the meanwhile a new client appears? Then you'd clear the domain with someone who uses it alive, which I think is a _really_ bad idea.

Comment (jajanusz) [reply src/schedule/ll_schedule.c]: Ah ok, so you need it for that case. Your are right then.

Comment (tlauda) [reply src/schedule/ll_schedule.c]: @paulstelian97 @jajanusz Yes, this is the exact problem. Thousands of multicore cycles showed that sometimes we could have race here.

**Comment by tlauda** [comment]
SOFCI TEST

**Comment by paulstelian97** [comment]
@tlauda I'm having trouble trying to rebase on top of this patchset. My current version of topic/imx_wip (which contains the incomplete ESAI driver as well as some testing code which shall not be submitted) works fine for the most part (minor issues). After rebasing on top of this PR I'm getting some interesting freezes on Stop (Xrun and the DSP becomes unresponsive to future IPCs). Is there something I need to do with my DMA drivers to fix that?

**Comment by tlauda** [comment]
@paulstelian97 It's weird. This patch just adds new domain (which I assume you don't even use) and mostly makes some topology changes. Other than that, just some cosmetics. Could you verify if reverting "pipeline: trigger pipeline in non-atomic context" helps?

**Comment by paulstelian97** [comment]
I will try that after regenerating the topology files (may that be a possibility as well?)

**Comment by paulstelian97** [comment]
Doesn't seem to help for some reason. And from what I have seen (I tried) I don't think I can rebase on top of individual commits without getting even worse breakage (to be able to identify the culprit). Any pointers on how I can check?

**Comment by tlauda** [comment]
The only thing that can potentially break anything is the last commit from this PR. It's small change, so just remove it manually. I don't see any other potential causes.

**Comment by paulstelian97** [comment]
Well reverting this one doesn't seem to help. So guess I'll have to directly compare the branches so I see what changes have been introduced and what could be the cause.

**Comment by paulstelian97** [comment]
b2836373 seems to also cause big trouble and I cannot rebase on top of it. Which is pretty funny. With this one added in (but none of the newer ones, and the other one with the timers reverted) I'm unable to make it work. Also tried to add an extra uint32_t field just to allow that to align and it still breaks.

**Comment by paulstelian97** [comment]
Removing the reads from dev->pipeline->ipc_pipe in params (for host and dai, not sure if I can just remove one of them, will test now) also makes it fine. So it's not an alignment issue.

**Comment by paulstelian97** [comment]
I need to know what already uses said period so I can investigate. Because while I can see it being valid in the print I get serious issues with actually using it.

**Comment by tlauda** [comment]
It's only used by DMA single channel domain, so it seems you have some kind of memory leaks, rather than functionality problems.

**Comment by paulstelian97** [comment]
I will try to rebase and revert each commit which causes trouble at least in part and let you know which changes actually cause me trouble.

**Comment by paulstelian97** [comment]
Alright, apparently something has suddenly become a lot more time-sensitive and our local hacks for traces (a serial console driver which operates in polling mode) causes IPCs to be dropped in such a way that everything breaks. Sorry for bothering. Even without reverting the commit you mentioned I still don't have issues, although I now have other issues (cannot figure out at all how to reliably get traces from the test board now that serial is no longer an option)

**Pull request merged**